src/rbdmap-generator

```diff
@@ -0,0 +1,282 @@
+                if a.startswith(SYSTEMD_KEY_PREFIX):
+                    if len(kv) < 2:
+                        logger.warning('option %s should contains a value, ignoring', a)
+                        continue
+                    k, v = kv
+                    k = k[len(SYSTEMD_KEY_PREFIX):]
+                    if k not in systemd_unit_options:
+                        logger.warning('unknown systemd option %s, ignoring', a)
+                        continue
+                    systemd_unit_options[k].append(v)
+                    continue
+            if kv[0] == 'options':
+                if len(kv) < 2:
+                    logger.warning('option %s should contains a value, ignoring', a)
+                    continue
+                rbd_map_arg_options += kv[1].split(',')
+                continue
+            if kv[0] in RBD_MAP_ARGS:
+                rbd_map_args.append(a)
+                if kv[0] == DEVICE_TYPE_KEY:
+                    if len(kv) < 2:
+                        raise ValueError('{} should contains a value'.format(DEVICE_TYPE_KEY))
+                    device_type = kv[1]
+                continue
+
+            rbd_map_arg_options.append(a)
+
+    if rbd_map_arg_options:
+        rbd_map_args.append('options=' + ','.join(rbd_map_arg_options))
+
+    if bool_options['auto']:
+        if bool_options['fail']:
+            systemd_unit_options['required-by'].append(RBDMAP_TARGET)
+        else:
+            systemd_unit_options['wanted-by'].append(RBDMAP_TARGET)
+    if bool_options['fail']:
+        systemd_unit_options['before'].append(RBDMAP_TARGET)
+
+    return systemd_unit_options, device_type, rbd_map_args
+
+def generate_one(line: str, dir_fd: int):
+    def opener(path, flags):
+        return os.open(path, flags, dir_fd=dir_fd, mode=0o644)
+
+    cells = line.split(maxsplit=1)
+    rbd_spec = cells[0]
+    map_opts = cells[1] if len(cells) >= 2 else None
+    rbd_spec = normalize_rbd_spec(rbd_spec)
+    rbd_dev_path = '/dev/rbd/' + rbd_spec
+    map_unit_name = 'rbdmap@{}.service'.format(systemd_escape(rbd_spec))
+    device_unit_name = systemd_escape_path(rbd_dev_path) + '.device'
+    systemd_unit_options, device_type, rbd_map_args = parse_opts(map_opts)
+
+    unit_extra = []
+    unit_options_map = [
+        ['requires', 'Requires'],
+        ['requires-mounts-for', 'RequiresMountsFor'],
+        ['before', 'Before'],
+        ['after', 'After'],
+    ]
+    for k, option in unit_options_map:
+        if systemd_unit_options[k]:
+            unit_extra.append(option + '=' + ' '.join(systemd_unit_options[k]))
+
+    service_type = 'Type=oneshot\nRemainAfterExit=yes'
+    if device_type == 'nbd':
+        # TODO: rbd-nbd --pid-file does not work https://github.com/ceph/ceph/pull/40844#issuecomment-820168209
+        # service_type = 'Type=forking\nPIDFile=%t/%n.pid'
+        # rbd_map_args.append('pid-file=%t/%n.pid')
+        service_type = 'Type=forking'
+    elif device_type not in ('krbd', None):
```

Comment:
Can you elaborate on what exactly do we lose by not having `PIDFile` (and hence not bothering with `--pid-file`)?  `rbd-nbd` is not a traditional service daemon after all, so I wonder what is the reasoning behind `Type=forking` instead of `Type=exec`?

Reply:
`Type=exec` is new in systemd 240, and is not available to Ubuntu 18.04. And systemd will not monitor the forked `rbd-nbd` process.

`forking` because systemd will invoke `rbd`, which will invoke `rbd-nbd`. Then `rbd` process will exit after the RBD device is successfully mapped. After that, systemd will consider the `rbdmap@.servce` unit active, and will monitor the `rbd-nbd`, change the unit state accordingly when `rbd-nbd` exit. Then units depending on it may react to the state change. `PIDFile=` is used to tell systemd which process it should monitor (the MainPID). And if it is not provided, systemd will guess one.

Reply:
Sorry, I meant `oneshot` -- the one that is used for `krbd`.

What I'm asking is if we even want systemd to monitor `rbd-nbd`.  Consider `krbd` case: there is nothing to monitor, if someone manually unmaps the image that has been mapped by the generated unit, systemd wouldn't know.  Is there a reason not to follow the same pattern for `rbd-nbd`?

Reply:
That's true, and `oneshot` will work. But in the nbd case, we have a process to monitor, and systemd can monitor it. Why not do it? And I suppose someone may query systemd for the MainPID in the future.

In the krbd case, I also considered adding a `BindsTo=dev-rbd-xxx-xxx.device` to the generated unit file. So when the device is manually unmapped, we can deactivate `rbdmap@.service` unit. But `crypttab` seems does not have this.

Reply:
One advantage of deactivating the unit properly is that systemd will try to map the RBD device again when the unit is pulled in again.

Reply:
We need to be consistent between `krbd` and `nbd`.  There is no daemon or PID in the krbd case.

The `BindTo` thing sounds like something that might work the same way for both `krbd` and `nbd`, but given that `cryptsetup` doesn't do it, we probably shouldn't either.  Let's keep it simple for now.

Reply:
IMO, there are some differences between `krbd` and `nbd`, and they are not entirely consistent naturally. `nbd` process can be killed (OOM, etc.), or crash (because of a bug, network issue, etc.). And administrators may decide to react when this does happen (e.g. by overriding the `RestartSec=`/`OnFailure=`). With `forking`, we definitely give local administrators (like me) more choices.

But if you insist, I will change it back to `simple`.

And just noticed, nbd will not create a symlink under `/dev/rbd/`, will it? Then although we can map the device, we cannot get mount in `/etc/fstab` to work, because we don't have a stable device path. Maybe we should delay the support for nbd.

Reply:
They are not naturally consistent, but for the purposes of the generated units which are really about calling `rbd device map` and not much else, I think it makes sense to have them behave the same.

Yes, no symlink would be created for `nbd`.  As far as the generator goes, I think there should be no special cases for `nbd` in the generator script.  So whether it is delayed or not, the code should be more or less the same.